scripts/upload_dataset.py

```diff
@@ -0,0 +1,193 @@
+        training_key = os.getenv("TRAINING_KEY")
+
+        credentials = ApiKeyCredentials(in_headers={"Training-key": training_key})
+        self.trainer = CustomVisionTrainingClient(ENDPOINT, credentials)
+
+        self.project_name = project_name
+
+        self.max_byte_size = 4000000
+
+    def _connect_to_or_create_project(self, project_name: str) -> Project:
+        projects = self.trainer.get_projects()
+        project_id = None
+        for project in projects:
+            if project.name == project_name:
+                project_id = project.id
+
+        if project_id is not None:
+            print("Connecting to existing project...")
+            return self.trainer.get_project(project_id)
+
+        print("Creating new project...")
+        obj_detection_domain = next(
+            domain
+            for domain in self.trainer.get_domains()
+            if domain.type == "ObjectDetection" and domain.name == "General"
+        )
+        return self.trainer.create_project(
+            project_name, domain_id=obj_detection_domain.id
+        )
+
+    def _get_or_create_tag(self, tag_name) -> Tag:
+        tags = self.trainer.get_tags(self.project.id)
+        for tag in tags:
+            if tag.name == tag_name:
+                return self.trainer.get_tag(self.project.id, tag.id)
+
+        return self.trainer.create_tag(self.project.id, tag_name)
+
+    def _read_and_resize_image(self, image_path: str) -> bytes:
+        image = cv2.imread(image_path, cv2.IMREAD_UNCHANGED)
+        current_byte_size = len(cv2.imencode(".jpg", image)[1].tobytes())
+        if current_byte_size > self.max_byte_size:
+            scale_factor: int = int(cbrt(self.max_byte_size / current_byte_size) * 95)
+            encode_param = [int(cv2.IMWRITE_JPEG_QUALITY), scale_factor]
+            image_bytes = cv2.imencode(".jpg", image, encode_param)[1].tobytes()
+        else:
+            image_bytes = cv2.imencode(".jpg", image)[1].tobytes()
+
+        return image_bytes
+
+    def _read_annotation_file(self, annotation_path: str) -> list:
+        annotations = []
+        with open(annotation_path, "r") as f:
+
+            for line in f:
+                line = line.strip()
+                parameter_list = line.split(" ")
+                label = int(parameter_list[0])
+                x, y, w, h = list(map(float, parameter_list[1:]))
+
+                left = x - w / 2
+                if left < 0:  # Accounting for previous rounding error
+                    left = 0
+                top = y - h / 2
+                if top < 0:  # Accounting for previous rounding error
+                    top = 0
+
+                if left + w > 1:  # Accounting for previous rounding error
+                    w = 1 - left
+                if top + h > 1:  # Accounting for previous rounding error
+                    h = 1 - top
+
+                if label == 0:
+                    tag_id = self.red_car_seal_tag.id
+                elif label == 1:
```

Comment:
Make this more general purpose and include the mapping between annotation labels and tags as parameter in the constructor?

Reply:
Agreed